pgdog/src/backend/pool/dns_cache.rs

```diff
@@ -0,0 +1,362 @@
+        Ok(ip)
+    }
+
+    /// Start the background refresh loop.
+    /// This spawns an infinite loop that runs until the process exits.
+    /// For testing, use short TTLs and allow the test runtime to handle shutdown.
+    pub fn start_refresh_loop(self: &Arc<Self>) {
+        let interval = Self::ttl();
+
+        let cache_ref = Arc::clone(self);
+        tokio::spawn(async move {
+            loop {
+                println!("-> I AM REFRESHING!!! {:?}", interval);
+                sleep(interval).await;
+                cache_ref.refresh_all_hostnames().await;
+            }
+        });
+    }
+}
+
+// -------------------------------------------------------------------------------------------------
+// ------ DnsCache :: Private methods --------------------------------------------------------------
+
+impl DnsCache {
+    /// Get the DNS refresh interval (short in tests, config-based otherwise).
+    fn ttl() -> Duration {
+        if cfg!(test) {
+            return Duration::from_millis(50);
+        }
+
+        config()
+            .config
+            .general
+            .dns_ttl()
+            .unwrap_or(Duration::from_secs(60)) // Should never happen in practice
+    }
+
+    /// Get IP address from cache only. Returns None if not cached or expired.
+    fn get_cached_ip(&self, hostname: &str) -> Option<IpAddr> {
+        if let Ok(ip) = hostname.parse::<IpAddr>() {
+            return Some(ip);
+        }
+
+        self.cache.read().get(hostname).copied()
+    }
+
+    /// Refresh all hostnames in the refresh list.
+    async fn refresh_all_hostnames(self: &Arc<Self>) {
+        let hostnames_to_refresh: Vec<String> = self.hostnames.read().iter().cloned().collect();
+
+        let tasks: Vec<_> = hostnames_to_refresh
+            .into_iter()
+            .map(|hostname| {
+                let cache_ref = Arc::clone(self);
+                tokio::spawn(async move {
+                    if let Err(e) = cache_ref.resolve_and_cache(&hostname).await {
+                        error!("failed to refresh DNS for {}: {}", hostname, e);
+                        println!("failed to refresh DNS for {}: {}", hostname, e);
+                    } else {
+                        println!("refreshed hostname: {}", hostname);
+                        debug!("refreshed hostname: {}", hostname);
+                    }
+                })
+            })
+            .collect();
+
+        join_all(tasks).await;
+    }
+
+    /// Do the actual DNS resolution and cache the result.
+    async fn resolve_and_cache(&self, hostname: &str) -> Result<IpAddr, Error> {
+        let response = self
+            .resolver
+            .lookup_ip(hostname)
+            .await
+            .map_err(|e: ResolveError| Error::DnsLookupError(format!("{}: {}", hostname, e)))?;
+
+        let ip = response
+            .iter()
+            .next()
+            .ok_or_else(|| Error::DnsLookupError(format!("{}: no IPs found", hostname)))?;
```

Comment:
The error enum expects the hostname here, not an error message. It would look pretty weird in the logs.